File: boxsvg/validation.py
"""Input validation for box requests."""

from __future__ import annotations

import os
from pathlib import Path

from boxsvg.models import (
    BoxRequest,
    SUPPORTED_MATERIALS,
    SUPPORTED_STYLES,
    SUPPORTED_UNITS,
)


class ValidationError(Exception):
    pass


def validate_request(request: BoxRequest) -> list[str]:
    """Validate a BoxRequest. Returns list of error messages (empty if valid)."""
    errors: list[str] = []

    if request.style not in SUPPORTED_STYLES:
        errors.append(
            f"Unsupported style: '{request.style}'. "
            f"Supported: {', '.join(SUPPORTED_STYLES)}"
        )

    if request.units not in SUPPORTED_UNITS:
        errors.append(
            f"Unsupported units: '{request.units}'. "
            f"Supported: {', '.join(SUPPORTED_UNITS)}"
        )

    if request.material not in SUPPORTED_MATERIALS:
        errors.append(
            f"Unsupported material: '{request.material}'. "
            f"Supported: {', '.join(SUPPORTED_MATERIALS)}"
        )

    for dim_name in ("length", "width", "height"):
        value = getattr(request, dim_name)
        if value <= 0:
            errors.append(f"{dim_name} must be greater than 0, got {value}")

    if request.thickness is not None and request.thickness <= 0:
        errors.append(f"thickness must be greater than 0, got {request.thickness}")

    if request.kerf is not None and request.kerf < 0:
        errors.append(f"kerf must be non-negative, got {request.kerf}")

    return errors


def validate_output_path(path: str, force: bool = False) -> str | None:
    """Validate output file path. Returns error message or None."""
    p = Path(path)
    if p.exists() and not force:
        return f"Output file '{path}' already exists. Use --force to overwrite."
    parent = p.parent
    if not parent.exists():
        return f"Output directory '{parent}' does not exist."
    if parent.exists() and not os.access(parent, os.W_OK):
        return f"Output directory '{parent}' is not writable."
    return None


def warn_unusual(request: BoxRequest) -> list[str]:
    """Return warnings for unusual but not invalid inputs."""
    warnings: list[str] = []
    max_dim = max(request.length, request.width, request.height)
    min_dim = min(request.length, request.width, request.height)
    if max_dim / min_dim > 10:
        warnings.append("Extreme aspect ratio detected — verify dimensions are correct.")
    if request.kerf is not None and request.kerf > 0.05:
        warnings.append(
            f"Kerf value {request.kerf} seems high — typical values are under 0.05 inches."
        )
    return warnings

File: web/src/core/validation.ts
import type { BoxRequest } from "./models";
import {
  SUPPORTED_STYLES,
  SUPPORTED_UNITS,
  SUPPORTED_MATERIALS,
  SUPPORTED_LID_FITS,
  getStyleDefinition,
} from "./models";

export function validateRequest(req: BoxRequest): string[] {
  const errors: string[] = [];

  if (!SUPPORTED_STYLES.includes(req.style)) {
    errors.push(`Unsupported style: '${req.style}'. Supported: ${SUPPORTED_STYLES.join(", ")}`);
  }
  if (!SUPPORTED_UNITS.includes(req.units)) {
    errors.push(`Unsupported units: '${req.units}'. Supported: ${SUPPORTED_UNITS.join(", ")}`);
  }
  if (!SUPPORTED_MATERIALS.includes(req.material)) {
    errors.push(`Unsupported material: '${req.material}'. Supported: ${SUPPORTED_MATERIALS.join(", ")}`);
  }
  if (!SUPPORTED_LID_FITS.includes(req.lid)) {
    errors.push(`Unsupported lid fit: '${req.lid}'. Supported: ${SUPPORTED_LID_FITS.join(", ")}`);
  }

  for (const dim of ["length", "width", "height"] as const) {
    if (req[dim] <= 0) {
      errors.push(`${dim} must be greater than 0, got ${req[dim]}`);
    }
  }

  if (req.thickness !== undefined && req.thickness <= 0) {
    errors.push(`thickness must be greater than 0, got ${req.thickness}`);
  }
  if (req.kerf !== undefined && req.kerf < 0) {
    errors.push(`kerf must be non-negative, got ${req.kerf}`);
  }

  if (req.height > req.length || req.height > req.width) {
    errors.push("Height should be the smallest dimension");
  }

  // Style-specific validation
  const styleDef = getStyleDefinition(req.style);
  if (styleDef?.validator) {
    errors.push(...styleDef.validator(req));
  }

  return errors;
}

export function warnUnusual(req: BoxRequest): string[] {
  const warnings: string[] = [];
  const maxDim = Math.max(req.length, req.width, req.height);
  const minDim = Math.min(req.length, req.width, req.height);

  if (maxDim / minDim > 10) {
    warnings.push("Extreme aspect ratio detected — verify dimensions are correct.");
  }
  if (req.kerf !== undefined && req.kerf > 0.05) {
    warnings.push(`Kerf value ${req.kerf} seems high — typical values are under 0.05 inches.`);
  }

  return warnings;
}
